sol-did/programs/sol-did/src/lib.rs

```diff
@@ -93,3 +96,318 @@ pub mod sol_did {
         instructions::migrate(ctx)
     }
 }
+
+// /// Given a DidAccount, and a list of controlling did accountInfos,
+// /// Parse the controlling did accountInfos into DidAccounts,
+// /// and validate that the controller chain is correct
+// fn valid_controller_chain<'a, 'b: 'a>(
+//     did_data: &'a Account<DidAccount>,
+//     did_accounts: &'b [AccountInfo<'b>],
+// ) -> Result<Vec<Account<'b, DidAccount>>> {
+//     let controller_chain: Vec<Account<DidAccount>> = did_accounts
+//         .iter()
+//         .map(Account::try_from)
+//         .collect::<Result<Vec<Account<DidAccount>>>>()?;
+//
+//     if !did_data.is_controlled_by(controller_chain.as_slice()) {
+//         return Err(error!(DidSolError::InvalidControllerChain));
+//     }
+//
+//     Ok(controller_chain)
+// }
+//
+// fn last_in_valid_controller_chain<'a, 'b: 'a>(
+//     did_data: &'a Account<DidAccount>,
+//     did_accounts: &'b [AccountInfo<'b>],
+// ) -> Result<Option<&'b Account<'b, DidAccount>>> {
+//     valid_controller_chain(did_data, did_accounts).map(|chain| chain.last())
+// }
+
+/// Will return if given solana public key, or ethereum address (as derived from an ethereum signature)
+/// is a valid authority (CAPABILITY_INVOCATION) on the given did_account.
+/// Authorities may be direct, or via a chain of controlling did accounts.
+/// In the latter case, the chain must be provided in the following order:
+/// did_account -> controlling_did_accounts[0] -> ... -> controlling_did_accounts[n] -> authority
+/// where '->' represents the relationship "is controlled by".
+pub fn is_authority(
+    did_account: &AccountInfo,
+    controlling_did_accounts: &[AccountInfo],
+    sol_authority: &Pubkey,
+    eth_message: &[u8],
+    eth_raw_signature: Option<&Secp256k1RawSignature>,
+    filter_fragment: Option<&String>,
+) -> Result<bool> {
+    if did_account.owner == &System::id() && did_account.lamports() == 0 {
```

Comment:
Why do we check for 0 lamports here? (and the owner actually)? Why don't we just assume NO data `did_account.data` means generative?